### PR TITLE
Enable C++/WinRT heap enforcement

### PR DESCRIPTION
**Pull request by vmoroz**
This is to address issue https://github.com/microsoft/react-native-windows/issues/3807.
Previously the ReactRootView was created using std::make_shared and I guess it was the main reason to disable the C++/WinRT check that all C++/WinRT objects are allocated on the heap instead of stack.

The ReactRootView behavior was changed in https://github.com/microsoft/react-native-windows/pull/4022 . We do not use the old ReactRootView from ReactUWP anymore and instead ReactRootView is a pure XAML control, while the other part that we named ReactRootControl is not C++/Winrt based.

In this change:
- We removed the C++/WinRT heap enforcement suppression from project files.
- Fixed code that stopped to compile due to the heap enforcement. These were most probably bugs that would show only for asynchronous code and would be difficult to debug.

@chrisglein, @licanhua could you please review the https://github.com/microsoft/react-native-windows/issues/4019 which was part of the bug being fixed. I believe the issue could be closed because we already use the C++/WinRT -optimized flag.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4045)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/ABICxxModule.cpp [hunk 1]
```diff
@@ -58,14 +58,14 @@ std::vector<CxxModule::Method> ABICxxModule::getMethods() {
 void ABICxxModule::InitEvents(std::vector<ABICxxModuleEventHandlerSetter> const &eventHandlerSetters) noexcept {
   for (auto &eventHandler : eventHandlerSetters) {
     eventHandler.EventHandlerSetter([ this, name = eventHandler.Name ](ReactArgWriter const &argWriter) noexcept {
-      DynamicWriter writer;
+      auto writer = make<DynamicWriter>();
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ABICxxModule.cpp]: >auto [](start = 6, length = 4)

(nit) why do we need this change?

File: vnext/Microsoft.ReactNative/ABICxxModule.cpp [hunk 2]
```diff
@@ -58,14 +58,14 @@ std::vector<CxxModule::Method> ABICxxModule::getMethods() {
...
       writer.WriteArrayBegin();
       writer.WriteString(winrt::to_hstring(name));
       argWriter(writer);
       writer.WriteArrayEnd();
 
       std::string emitterName = m_eventEmitterName.empty() ? DefaultEventEmitterName : m_eventEmitterName;
-      m_reactContext->CallJSFunction(std::move(emitterName), "emit", writer.TakeValue());
+      m_reactContext->CallJSFunction(std::move(emitterName), "emit", get_self<DynamicWriter>(writer)->TakeValue());
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/ABICxxModule.cpp]: >get_self [](start = 69, length = 8)

why is this needed? doesn't operator-> deref the implementation?

File: vnext/Microsoft.ReactNative/ABICxxModule.cpp (hunk 1 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/ABICxxModule.cpp]: It seems that previously this code was creating classes on the call stack and the new check causes a compilation error. Our tests are passing because they do not asynchronous code inside. Though customer code would fail if they ever do.

---
In reply to: [376125176](https://github.com/microsoft/react-native-windows/pull/4045#discussion_r376125176) [](ancestors = 376125176)

File: vnext/Microsoft.ReactNative/ABICxxModule.cpp (hunk 2 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/ABICxxModule.cpp]: No, the writer here is winrt::com_ref<IJSValueWriter>. To get from the interface to the public method of the class, we need to use the get_self to acquire the implementation class first. See for more details: [https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/author-apis](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/author-apis)

---
In reply to: [376139460](https://github.com/microsoft/react-native-windows/pull/4045#discussion_r376139460) [](ancestors = 376139460)
